File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.java
class Solution {
    public int[][] generateMatrix(int n) {

        int[][] ans = new int[n][n];

        int top = 0, left = 0;
        int right = n - 1, bottom = n - 1;

        int num = 1;

        while (num <= n * n) {

            // left → right
            for (int i = left; i <= right; i++) {
                ans[top][i] = num++;
            }
            top++;

            // top → bottom
            for (int i = top; i <= bottom; i++) {
                ans[i][right] = num++;
            }
            right--;

            // right → left
            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    ans[bottom][i] = num++;
                }
                bottom--;
            }

            // bottom → top
            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    ans[i][left] = num++;
                }
                left++;
            }
        }

        return ans;
    }
}

File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> ans(n , vector<int>(n));
        int top = 0 , bottom = n-1 , left = 0 , right = n-1 , index = 1;
        while(top <= bottom && left <= right){
            for(int i = left ; i <= right ; i++){
                ans[top][i] = index++ ;

            }
            top++;
            for(int i = top ; i <= bottom ; i++){
                ans[i][right] = index++;

            }
            right--;
            if(top <= bottom){
                for(int i = right ; i >= left ; i--){
                    ans[bottom][i] = index++;

                }
                bottom--;
            }
            if(left <= right){
                for(int i = bottom ; i >= top ; i--){
                    ans[i][left] = index++;

                }
                left++;
            }
        }
        return ans;
    }
};